### PR TITLE
Develop wiregrid encoder

### PR DESCRIPTION
**Pull request by sadachi5**
## Description

Make a new agent: wiregrid_encoder
This agent is run for wire grid loader.
It records the rotary encoder data sent by beagleboneblack in the electronics for wire grid loader.

## Motivation and Context

Add the new agent to the develop branch

## How Has This Been Tested?

This agent has been tested in a test bench of the wire grid loader at Kyoto University.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [x] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

### CODE REVIEW

File: agents/wiregrid_encoder/signal_parser.py
Comment (BrianJKoopman) [inline agents/wiregrid_encoder/signal_parser.py:60]: You don't really need this else and can get rid of it and un-indent the below section. If the above if statement is entered you'll hit that `break`.

File: agents/wiregrid_encoder/wiregrid_encoder.py
Comment (BrianJKoopman) [inline agents/wiregrid_encoder/wiregrid_encoder.py:102]: Since I'm not familiar with the external script on the beaglebone -- when are these two cases considered? Is it in IRIG mode or Enocder mode? Can both modes be active at once?

If both modes can be active, then these should be combined, else a client expecting the Encoder mode might be the IRIG mode response and not know what to do.

In any case, you should add a timestamp to these so that the client knows if the data is stale or not.

File: agents/wiregrid_encoder/wiregrid_encoder.py
Comment (BrianJKoopman) [inline agents/wiregrid_encoder/wiregrid_encoder.py:317]: Does running this flush_buffer cause issues?

**Review by BrianJKoopman — COMMENTED** [review]
The Agent here is looking good, a couple small comments.

I also have a few questions about things, which may require some changes depending on the answers.

File: docs/agents/wiregrid_encoder.rst
Comment (sadachi5) [reply docs/agents/wiregrid_encoder.rst:27]: I added the github URL (simonsobs/wire_grid_loader).

File: docs/agents/wiregrid_encoder.rst
Comment (sadachi5) [reply docs/agents/wiregrid_encoder.rst:87]: Actually, both of them will not be changed as far as the hardware, e.g. the encoder model, is not changed.
I think such kind of change is unexpected.
Anyway, this information is not necessary for the agent user.
So if you want, I can remove these sentences.

File: agents/wiregrid_encoder/wiregrid_encoder.py
Comment (sadachi5) [reply agents/wiregrid_encoder/wiregrid_encoder.py:137]: Done

File: agents/wiregrid_encoder/signal_parser.py
Comment (sadachi5) [reply agents/wiregrid_encoder/signal_parser.py:60]: Done

File: agents/wiregrid_encoder/signal_parser.py
Comment (sadachi5) [reply agents/wiregrid_encoder/signal_parser.py:100]: done

File: docs/agents/wiregrid_encoder.rst
Comment (sadachi5) [reply docs/agents/wiregrid_encoder.rst:1]: done

File: agents/wiregrid_encoder/wiregrid_encoder.py
Comment (sadachi5) [reply agents/wiregrid_encoder/wiregrid_encoder.py:102]: I forgot to write timestamp in the docstring, but timestamp data is already implemented.

The beaglebone sends the IRIG data and Encoder data randomly. 
So, this agent record the IRIG and encoder data in one agent in one agent.
I mean an example of the list of session data is:
[{timestamp, IRIG data}, {timestamp, Encoder data}, {timestamp, Encoder data}, {timestamp, Encoder data}{timestamp, IRIG data}, ...]

File: docs/agents/wiregrid_encoder.rst
Comment (BrianJKoopman) [reply docs/agents/wiregrid_encoder.rst:87]: I think it's fine to include, but maybe add here that these values will rarely be changed.

File: docs/agents/wiregrid_encoder.rst
Comment (sadachi5) [reply docs/agents/wiregrid_encoder.rst:80]: I removed these lines.

File: docs/agents/wiregrid_encoder.rst
Comment (sadachi5) [reply docs/agents/wiregrid_encoder.rst]: I added it.

File: agents/wiregrid_encoder/wiregrid_encoder.py
Comment (sadachi5) [reply agents/wiregrid_encoder/wiregrid_encoder.py:61]: The buffering occurs at ~200Hz.
The full data has 1000 points data and the rough data has 10 points.
So the encoder data are obtained at 200kHz and 2kHz in the full and rough data, respectively.

Comment (BrianJKoopman) [reply agents/wiregrid_encoder/wiregrid_encoder.py:61]: Sorry, just want to make sure I'm following. The final data rate written to disk/influxdb is then 1kHz and 10 Hz, respectively?

Comment (sadachi5) [reply agents/wiregrid_encoder/wiregrid_encoder.py:61]: Sorry for confusing you.
The final data rate written to disk/influxdb is the same, which is ~200Hz (I mentioned it as a buffering procedure).
There is a difference in the data size in each writing between the full data (1000 points) and rough data (10 points).

Comment (sadachi5) [reply agents/wiregrid_encoder/wiregrid_encoder.py:62]: I'm checking the rate with Hironobu. Please wait for a whille.

File: agents/wiregrid_encoder/wiregrid_encoder.py
Comment (sadachi5) [reply agents/wiregrid_encoder/wiregrid_encoder.py]: I merged the irig and encoder data as you suggested.
Please check the new version.

File: agents/wiregrid_encoder/wiregrid_encoder.py
Comment (sadachi5) [reply agents/wiregrid_encoder/wiregrid_encoder.py]: In the past, I and Hironobu faced the problem of increase of memory usage in this agent.
We removed it to avoid it just in case the flush_buffer caused some problem.
But, I don't know if it really causes it or not.

I think you know this problem a bit because Hironobu asked you about it.

Now, temporarily I added this flush_buffer() to check if it really makes any problem or not again.

File: agents/wiregrid_encoder/wiregrid_encoder.py
Comment (sadachi5) [reply agents/wiregrid_encoder/wiregrid_encoder.py:62]: Please forget the previous my statements about the rates.

According to Hironobu, the publish_buffer() (writing to disk/influxdb) occurs at 2Hz. 
The full data has 1000 data points. So the recorded encoder full data is 2kHz.

By the way, the sending data containing 100 data points from BBB to PC occurs at 20 Hz.

Comment (BrianJKoopman) [reply agents/wiregrid_encoder/wiregrid_encoder.py:61]: Great, thanks for checking the numbers. Those data rates to InfluxDB/g3 file sound alright to me.

File: agents/wiregrid_encoder/wiregrid_encoder.py
Comment (BrianJKoopman) [reply agents/wiregrid_encoder/wiregrid_encoder.py]: Gotcha, let me know how it goes, I won't let it hold up the PR, we can change things later if it proves to be an issue.

File: agents/wiregrid_encoder/wiregrid_encoder.py
Comment (BrianJKoopman) [reply agents/wiregrid_encoder/wiregrid_encoder.py]: Looks good to me, thanks for merging that.

**Review by BrianJKoopman — APPROVED** [review]
Thanks for the updates to this! I'm approving. I am, however, going to clean up the git history a bit (the extraction from the other PR left a lot of identical commits) and force push to this branch. After that I'll handle a merge of the latest develop branch, and will run the new pre-commit hooks, which are going to do a bit of auto-formatting. I might leave actually merging to develop until next week, but I think this is in a good place and can be merged.

**Pull request merged**
